### ISSUE TITLE
Server-side Chakra optimization  

### ISSUE THREAD

**Issue by markb-avalara** [Microsoft/ChakraCore#4406]
I've got a .NET service that runs small snippits of JavaScript. It needs to be super fast so I've pre-allocated a pool of chakra runtimes. I'm attempting to re-use the runtimes but after running a single threaded stress for a bit I encounter "Wrong thread" exception when attempting to set a newly created context on the runtime.  I can clearly see from debug messages that the previous use of the runtime had completed.

Is this the best way to approach this problem? When attempting to run several scripts at the same time and keep them completely isolated is my idea of a pool of re-usable runtimes and new context for each request correct? What is the lifetime of a context? When does it get cleaned up? I'm also seeing out of memory exceptions from Chakra.

Should I just be creating brand new runtimes for each request?

**Comment by liminzhu** [comment]
> Wrong thread

Was the runtime on a previous thread? If so, you need to `JsSetCurrentContext(JS_INVALID_REFERENCE)` on that thread to give up the runtime.

> When attempting to run several scripts at the same time and keep them completely isolated is my idea of a pool of re-usable runtimes and new context for each request correct? 

This is the right idea. A new context is a new execution environment for a new incoming request.

> What is the lifetime of a context? When does it get cleaned up? I'm also seeing out of memory exceptions from Chakra.

Contexts are susceptible to GC like regular objects.It's pinned when set as current context, otherwise GC can clean it up anytime unless `JsAddRef` is called.

**Comment by markb-avalara** [comment]
> Was the runtime previously on a different thread? If so, you need to JsSetCurrentContext(JS_INVALID_REFERENCE) on that thread to give up the runtime

Yes, probably a different thread. In order to terminate scripts that run too long I run the script using a Task so the script is running on some arbitrary thread from the thread pool. And I'm sure that at the end of execution that the context is set to JS_INVALID_REFERENCE because I'm using the Scope object in the C# wrapper that comes with the samples. In the Dispose of that object it sets context to previous value which would be JS_INVALID_REFERENCE.

**Comment by liminzhu** [comment]
I see. Can you send us a small repro? We can take a closer look at what happened.

**Comment by markb-avalara** [comment]
I've attached a demo of the issue. It occurs in the timeout case where the script runs too long. Specifically when I attempt to re-enabled the runtime after I've terminated it at line 166 in Program.cs. Thanks for the help!
[ChakraRuntimePoolStress.zip](https://github.com/Microsoft/ChakraCore/files/1564127/ChakraRuntimePoolStress.zip)

**Comment by markb-avalara** [comment]
In the process of creating an app that reproduced the problem I found my Wrong Thread issue. I had a race condition in my async thread where a thread signal was inside the using statement for the Scope. So sometimes I'd attempt to renabled a runtime before the previous thread was finished. However, I fixed that and now encounter the second problem I've occasionally seen. The application just exits or I get a JavaScriptFatalException. I'm probably doing something wrong but I don't see it. Thanks for the help.

Run to encounter the issue. Doesn't take very long. Thanks again.

[ChakraRuntimePoolStress.zip](https://github.com/Microsoft/ChakraCore/files/1564316/ChakraRuntimePoolStress.zip)

**Comment by markb-avalara** [comment]
I attached a simpler app in the previous comment that demonstrates the issue. We won't be able to add Chakra to our production servers until we understand why it crashes in the timeout case.

**Comment by markb-avalara** [comment]
I'm closing this and reopening as another issue because this is marked as a question that has been answered.
